### PR TITLE
Render mermaid codeblocks

### PR DESCRIPTION
**Pull request by ben-krieger**
This PR adds support for rendering mermaid in codeblocks / code fences.

Examples I was able to find for this feature, like https://anis.se/posts/add-mermaidjs-support-to-hugo/, had a few issues.

1. Imported mermaid more than once
2. Imported mermaid from different sources/versions
3. Did not support dark mode

Sorry for not creating an issue first, like it recommends in CONTRIBUTING.md. However, I figured that this probably wasn't something that would be desired for the mainline release, since it changes existing behavior. If I'm wrong, maybe it can be put behind a config flag.

I also inserted a personal fix for https://github.com/mermaid-js/mermaid/issues/2481, which may need to be omitted if this PR is to be considered for merge.

Maintainers: feel free to close this PR if you're not interested in the feature. I just hope that it helps some poor fool like past me who searches for "geekdoc mermaid codeblock / code fence support!"

EDIT: My branch is currently based off of v0.33.2, because Mermaid v9.1.4 is broken in such a way that having multiple sequence diagrams in one page cumulatively joins them! So sequence 1, sequence 1+2, sequence 1+3.

EDIT 2: Mermaid v9.1.5 and v9.1.6 seem to fix the error, so I will rebase this to merge into main, which already has updated to v9.1.5.

EDIT 3: The mermaid 9.1.4 bug being referenced was https://github.com/mermaid-js/mermaid/issues/3305 and was fixed here: https://github.com/mermaid-js/mermaid/pull/3310

### CODE REVIEW

**Comment by xoxys** [comment]
Thanks for your effort. I would prefer an implementation using [rendering hooks](https://gohugo.io/templates/render-hooks/) as is approach can be used for other extensions e.g. KaTeX as well. This way, it should be also possible to re-use the same code from the existing shortcode.

**Comment by ben-krieger** [comment]
> Thanks for your effort. I would prefer an implementation using [rendering hooks](https://gohugo.io/templates/render-hooks/) as this approach can be used for other extensions e.g. KaTeX as well. This way, it should be also possible to re-use the same code from the existing shortcode.

I didn't realize you could do that for codeblocks now in Hugo! I'll take a look.

**Comment by ben-krieger** [comment]
And the solution is literally given in the docs... https://gohugo.io/content-management/diagrams/#mermaid-diagrams

But there is a small amount of work needed to integrate with the existing mermaid shortcode so that using a codeblock and a shortcode in the same page doesn't result in importing and initializing mermaid twice.

**Comment by xoxys** [comment]
The shortcode has already a check using the exact same code in the render template should avoid double imoorts.

**Comment by ben-krieger** [comment]
> The shortcode has already a check using the exact same code in the render template should avoid double imoorts.

I just mean change `.Page.Store.Get("hasMermaid")` from the doc example to match `.Page.Scratch.Get("mermaid")` from the theme source (and matching Set commands). :)

**Comment by ben-krieger** [comment]
I removed `mermaid.init` call which specified the two possible mermaid class names (`mermaid` and `language-mermaid`) and added rendering hook for a mermaid codeblock instead. I also rebased onto main.

File: src/js/mermaid.js [hunk 1]
```diff
@@ -23,6 +23,12 @@ document.addEventListener("DOMContentLoaded", function (event) {
         theme: theme,
         themeVariables: {
           darkMode: darkMode
+        },
+        mermaid: {
+          // Fix height of mermaid SVG elements (see https://github.com/mermaid-js/mermaid/issues/2481)
+          callback: (id) => {
+            document.getElementById(id).setAttribute("height", "100%")
+          }
```
Comment (ben-krieger) [inline src/js/mermaid.js:31]: Should I move this change to a separate PR?

File: layouts/_default/_markup/render-codeblock-mermaid.html [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  <!-- Include mermaid only first time -->
+  <script defer src="{{ index (index .Site.Data.assets "mermaid.js") "src" | relURL }}"></script>
+  {{ .Page.Scratch.Set "mermaid" true }}
+{{ end }}
+<!-- prettier-ignore-end -->
+
+{{ .Page.Scratch.Set "mermaid" true }}
+<pre class="gdoc-mermaid mermaid">
+  {{- .Inner -}}
+</pre>
```
Comment (ben-krieger) [inline layouts/_default/_markup/render-codeblock-mermaid.html:12]: Is there somewhere in the docs to specify a minimum Hugo version? This requires v0.93.0 or later, which was only released at the end of February 2022.

Comment (xoxys) [reply layouts/_default/_markup/render-codeblock-mermaid.html:12]: Yes there are three files to change:

- [README](https://github.com/thegeeklab/hugo-geekdoc/blob/main/README.md) Line 4
- https://github.com/thegeeklab/hugo-geekdoc/blob/main/exampleSite/content/en/_index.md Line 12
- https://github.com/thegeeklab/hugo-geekdoc/blob/main/theme.toml#L8

File: src/js/mermaid.js (hunk 1 above)
Comment (xoxys) [reply src/js/mermaid.js:31]: Yes, please create a dedicated PR, make it easier in case we need to revert it. Wondering why this was not fixed upstream yet if a fix would be that simple :thinking:

File: README.md [hunk 3]
```diff
@@ -22,6 +22,12 @@ npm install
 
 # run the build script to build required assets
 npm run build
+
```
Comment (xoxys) [inline README.md:25]: I would like to move this to a dedicated PR as well and implement it as a reusable NPM script. It also needs to be documented in the docs as well.

**Comment by xoxys** [comment]
Superseded by: https://github.com/thegeeklab/hugo-geekdoc/pull/485 https://github.com/thegeeklab/hugo-geekdoc/pull/486
